src/pdb/string.rs

```diff
@@ -0,0 +1,275 @@
+        // https://isrc.ifpi.org/downloads/ISRC_Bulletin-2015-01.pdf
+        if string.len() != 12 || !only_contains_ascii(&string) {
+            return Err(StringError::InvalidISRC);
+        }
+        Ok(Self(DeviceSQLStringImpl::Long {
+            content: LongBody::Isrc(NullString::from_string(string)),
+        }))
+    }
+    /// Extract the Rust string from the DeviceSQLString.
+    ///
+    /// Consumes itself in the process.
+    pub fn into_string(self) -> Result<String, StringError> {
+        match self.0 {
+            DeviceSQLStringImpl::ShortASCII { content: vec, .. }
+            | DeviceSQLStringImpl::Long {
+                content: LongBody::Ascii(vec),
+                ..
+            } => String::from_utf8(vec).map_err(|_| StringError::Encoding),
+            DeviceSQLStringImpl::Long {
+                content: LongBody::Isrc(str),
+                ..
+            } => str
+                .into_string_lossless()
+                .map_err(|_| StringError::Encoding),
+            DeviceSQLStringImpl::Long {
+                content: LongBody::Ucs2le(vec),
+            } => String::from_utf16(&vec).map_err(|_| StringError::Encoding),
+        }
+    }
+    /// Create an empty DeviceSQLString
+    ///
+    /// Should be used to construct known empty strings.
+    #[must_use]
+    pub const fn empty() -> Self {
+        Self(DeviceSQLStringImpl::ShortASCII {
+            content: Vec::new(),
+        })
+    }
+}
+
+/// A String encapsulating how DeviceSQL VARCHAR's are structured
+///
+/// This implementation forces them to be immutable for now
+#[binrw]
+#[derive(Debug, PartialEq)]
+enum DeviceSQLStringImpl {
+    /// Short-String optimization case
+    ShortASCII {
+        #[br(temp, assert(header & 0b1 == 1))]
+        #[bw(calc = (((content.len() + 1) << 1) | 1) as u8)]
+        header: u8,
+
+        #[br(count = (header >> 1) - 1)]
+        content: Vec<u8>,
+    },
+    /// Regular long form strings, containing possibly different encodings
+    Long {
+        #[br(temp)]
+        #[br(assert(flags & 0b1 == 0))]
+        #[bw(calc = content.flags())]
+        flags: u8,
+
+        #[br(temp)]
+        #[bw(calc = content.byte_count().unwrap() + 4)]
+        length: u16,
+
+        #[brw(magic(0u8))] // padding
+        #[br(args(flags, length - 4))]
+        content: LongBody,
+    },
+}
+
+#[binrw]
+#[derive(Debug, PartialEq)]
+#[br(import(flags: u8, len: u16))]
+enum LongBody {
+    // Ordering is important otherwise, UCS2LE strings could be parsed
+    // instead of the stricter ISRC
+
+    // ISRC strings are a bug/flaw in pioneers implementation, this is technically
+    // a semi-dirty workaround
+    #[br(pre_assert(flags == 0x90))]
+    Isrc(#[brw(magic = 0x3u8)] binrw::NullString),
+    #[br(pre_assert(flags == 0x40))]
+    Ascii(#[br(count = len)] Vec<u8>),
+    #[br(pre_assert(flags == 0x90))]
+    #[br(pre_assert(len % 2 == 0))]
+    Ucs2le(#[br(count = len / 2)] Vec<u16>),
```

Comment:
These enum members should be uppercase IMHO. 

```suggestion
    #[br(pre_assert(flags == 0x90))]
    ISRC(#[brw(magic = 0x3u8)] binrw::NullString),
    #[br(pre_assert(flags == 0x40))]
    ASCII(#[br(count = len)] Vec<u8>),
    #[br(pre_assert(flags == 0x90))]
    #[br(pre_assert(len % 2 == 0))]
    UCS2LE(#[br(count = len / 2)] Vec<u16>),
```

Reply:
`clippy::all` implies `clippy::upper_case_acronyms` see https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

We can make an exception but don't think we should because this is a style choice that opposes the established convention.

Reply:
ok